### ISSUE TITLE
Nix eval performance

### ISSUE THREAD

**Issue by infinisil** [tweag/nix-hour#74]
Let's investigate Nix evaluation performance! We could cover things like:
- https://nixos.wiki/wiki/Nix_Evaluation_Performance (written by me!)
- Laziness and thunks (also see https://github.com/tweag/nix-hour/issues/46)
- https://github.com/infinisil/fastnixarray which I wrote to solve [AoC 2020 Day 15](https://adventofcode.com/2020/day/15) [efficiently](https://github.com/infinisil/aoc20/blob/master/15/default.nix)

We will focus on this topic this week at Wednesday 15:00 CEST! If anybody has any extra questions about this topic, either:

- Ask them in this issue in advance with a comment
- Join the [livestream]() and use the chat to ask questions live

**Comment by asymmetric** [comment]
Is IFD in scope for this? Since it has an impact on evaluation performance.

**Comment by infinisil** [comment]
@asymmetric Good point, sounds in scope to me!

**Comment by tcurdt** [comment]
Interesting topic! Let's say I want to manage 100-1000 servers - how does this scale with NixOS?
Even with just 3 servers in my flake it feels like a drag sometimes.

**Comment by iwanb** [comment]
Not directly eval performance, but I'd be interested to know more how nix can avoid recomputing things which are already in the store, or in a substituter. E.g. I expect fixed output derivations or content-addressed derivations 'prune' the evaluation that needs to happen if they're already in the store?

And somewhat related, how does the 'toFile' built-in impact evaluation

**Comment by tcurdt** [comment]
> Not directly eval performance, but I'd be interested to know more how nix can avoid recomputing things which are already in the store

Yeah, that's a good one. I am sometimes puzzled about how much recomputing it takes to change (seemingly?) little things. Would be good to know how to "debug" this.
